overlord/devicestate/devicemgr.go

```diff
@@ -0,0 +1,384 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package devicestate implements the manager and state aspects responsible
+// for the device identity and policies.
+package devicestate
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"golang.org/x/crypto/openpgp/packet"
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+)
+
+// DeviceManager is responsible for managing the device identity and device
+// policies.
+type DeviceManager struct {
+	state  *state.State
+	runner *state.TaskRunner
+}
+
+// Manager returns a new device manager.
+func Manager(s *state.State) (*DeviceManager, error) {
+	runner := state.NewTaskRunner(s)
+
+	runner.AddHandler("generate-device-key", doGenerateDeviceKey, nil)
+	runner.AddHandler("deliver-serial-request", doDeliverSerialRequest, nil)
+	runner.AddHandler("retrieve-n-install-serial", doRetrieveAndInstallSerial, nil)
+
+	return &DeviceManager{state: s, runner: runner}, nil
+}
+
+func (m *DeviceManager) ensureOperational() error {
+	m.state.Lock()
+	defer m.state.Unlock()
+	// XXX: auth.Device/SetDevice should probably move to devicestate
+	// they are not quite just about auth and also we risk circular imports
```

Comment:
Although, we need to think about how AuthContext works there. The current design might end up cleaner for that reason. Happy to reconsider, though.